libtransmission/completion.cc

```diff
@@ -119,7 +119,7 @@ tr_completeness tr_completion::status() const
         return TR_SEED;
     }
 
-    if (size_now_ == sizeWhenDone())
+    if (size_now_ == sizeWhenDone() && sizeWhenDone() != 0)
```

Comment:
```suggestion
    if (hasMetainfo() && size_now_ == sizeWhenDone())
```

Reply:
`size_now_ == sizeWhenDone()` both are 0 for a magnet link

Reply:
also `hasMetainfo()` is returning true for the magnet links

Reply:
@ckerr
```
if (!hasNone() && size_now_ == sizeWhenDone())
```

works in this situation

Reply:
`hasMetainfo()` returns true iff we have any piece data for that torrent (tested by checking `!std::empty(blocks_)`. Sounds like either a bug in `hasMetainfo()` or a mistake in testing it here?

Reply:
yup,
for newly added magnet links `hasMetainfo()` returns `true` - at least on macOS
`hasMetadata()` on the other hand returns false, but this is not available in `completion.cc`

Reply:
how about something like this for `hasMetainfo()`:
```
[[nodiscard]] constexpr bool hasMetainfo() const
    {
        // magnet links
        if (size_when_done_ == 0) {
            return false;
        }
        return !std::empty(blocks_);
    }
```

Reply:
...I am not sure how that could be happening. There are no blocks for it to have, and it doesn't know what the block size of the torrent is, so at this point the count should always be zero. I hear what you're saying about the workaround but it smells like there's a bug here and I'd like to figure out if there's something else that needs to be fixed, too.

I added this test patch locally:

```diff
diff --git a/libtransmission/completion.h b/libtransmission/completion.h
index 522163be0..81f68b253 100644
--- a/libtransmission/completion.h
+++ b/libtransmission/completion.h
@@ -15,6 +15,8 @@
 #include <optional>
 #include <vector>
 
+#include <iostream> // nocommit
+
 #include "transmission.h"
 
 #include "block-info.h"
@@ -130,9 +132,11 @@ struct tr_completion
     [[nodiscard]] uint64_t countHasBytesInSpan(tr_byte_span_t) const;
 
 private:
-    [[nodiscard]] constexpr bool hasMetainfo() const
+    [[nodiscard]] bool hasMetainfo() const
     {
-        return !std::empty(blocks_);
+        auto const ret = !std::empty(blocks_);
+        std::cerr << __FILE__ << ':' << __LINE__ << ' ' << ret << std::endl;
+        return ret;
     }
 
     [[nodiscard]] uint64_t computeHasValid() const;
```

and then fired up an empty (no torrents) session and added a magnet link. What I got was:

```sh
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
/tmp/foo/transmission/libtransmission/completion.h:138 0
```

Could you describe the steps I should follow to try and test this to get this to return 1 instead of 0?

Reply:
strange,
I emptied my session, and started with new magnet links... and now I get the expected results...

Im not sure what's going on...
possibly cruft from an earlier build?